File: 1. Offline 1/Bank.cpp
#include <bits/stdc++.h>
#include "Account.cpp"
using namespace std;

#define ACCOUNT 23
#define EMPLOYEE 24
#define NOONE 25

vector<double> interestRates{0.1, 0.05, 0.15};
class Bank {
    int MD, officer[2], cashiers[5];
    vector<Account> accounts;
    stack<Account> LoanRequests;
    int yearOfOperation;
    double internalFund;
    
public:
    Bank() {
        cout << "Bank Created; MD, S1, S2, ";
        cout << "C1, C2, C3, C4, C5 created\n";

        internalFund = 1000000;
        yearOfOperation = 0;
    }

    double getInternalFund() {
        return internalFund;
    }

    void DecreaseInternalFund(double amount) {
        internalFund -= amount;
    }

    void increaseTime() {
        yearOfOperation++;

        //decrease loan
        for (Account account : accounts) {
            account.DecreaseLoan(account.getLoan()*0.1);
            account.IncreaseBalance(account.getBalance()*interestRates[account.getType()]);
            account.DecreaseBalance(500 < account.getBalance() ? 500 : account.getBalance());
        }
    }

    int CheckIfAccountExists(string username) {
        for (Account account : accounts) {
            if (account.getUsername() == username) {
                return ACCOUNT;
            }
        }
        return EMPLOYEE;
    }

    void addAccount(Account account) {
        accounts.push_back(account);
    }

    void addLoanRequest(Account account) {
        LoanRequests.push(account);
    }

    void removeLoanRequest() {
        LoanRequests.pop();
    }

    vector<Account> getAccounts() {
        return accounts;
    }

    int getTime() {
        return yearOfOperation;
    }


    Account AccountLogin(string username) {
        Account account;
        for (Account account : accounts) {
            if (account.getUsername() == username) {
                account.MakeActive();
                return account;
            }
        }
        return account;
        // whoIsActiveStatus = ACCOUNT;
    }

    void AccountLogout(string username) {
        for (Account account : accounts) {
            if (account.getUsername() == username) {
                account.MakeInactive();
            }
        }
    }
};

File: 1. Offline 1/Bank.java
import java.util.ArrayList;

public class Bank {
    double interestRates[] = {0.1, 0.05, 0.15}; //have to implement setInterestRates

    protected ArrayList<Account> accounts;
    protected ArrayList<Account> loanRequests;
    protected int year;
    protected double internalFund;

    Bank() {
//        System.out.println("Bank Created; MD, S1, S2, C1, C2, C3, C4, C5 created");
        accounts = new ArrayList<>();
        loanRequests = new ArrayList<>();
        internalFund = 1000000;
        year = 0;
    }

    public ArrayList<Account> getAccounts() {
        return accounts;
    }

    public int getYear() {
        return year;
    }

    public double getInternalFund() {
        return internalFund;
    }

    public void DecreaseInternalFund(double amount) {
        internalFund -= amount;
    }

    public void IncreaseYear() {
        year++;
        System.out.println("1 year passed");

        //decrease loans
        for (Account account : accounts) {
            account.DecreaseLoan(account.getLoan()*0.1);
            account.IncreaseBalance(account.getBalance()*interestRates[account.getTypeIndex()]);
            account.DecreaseBalance(500 < account.getBalance() ? 500 : account.getBalance());
        }
    }

    Account CheckIfAccountExists(String username) {
        for (Account account : accounts) {
            if (account.getUsername().equalsIgnoreCase(username)) {
                return account;
            }
        }
        return null;
    }

    boolean CheckIfValidLoanRequest(double amount) {
        if (amount < internalFund) {
            return true;
        }
        return false;
    }

    boolean CheckIfLoanRequestsPending() {
        return !loanRequests.isEmpty();
    }

    public void AddAccount(Account account) {
        accounts.add(account);
    }

    public void AddLoanRequest(Account account) {
        loanRequests.add(account);
    }

}
